### DISCUSSION TITLE
Access REMOTE_USER

### DISCUSSION THREAD

**Discussion by userg3003** [abersheeran/a2wsgi#11] [Ideas]
The CGI RFC provides for a REMOTE_USER variable, which holds the username of the user who has been authenticated by the front-end Web server. For a WSGI application, this can be obtained as environ["REMOTE_USER"], since a WSGI application gets access to all CGI variables. It would be nice if an ASGI application could get access to the same thing.
For example, if I run the Flask application under the Apache server, with authorization through Active Directory, I can get the name of the authorized user in the remote_user variable.
Note: missing variables (such as REMOTE_USER when no authentication has occurred) should be left out of the environ dictionary (
[PEP 3333](https://www.python.org/dev/peps/pep-3333/#environ-variables)).

To add the ability to access REMOTE_USER , you can do the following in asgi.py (in def build_scope(environ: Environ) -> Scope:):
```python
def build_scope(environ: Environ) -> Scope:

    ...

    if environ.get("REMOTE_USER"):
        remote_user = environ.get("REMOTE_USER")
    else:
        remote_user = None


    return {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "3.0"},
        "http_version": environ.get("SERVER_PROTOCOL", "http/1.0").split("/")[1],
        "method": environ["REQUEST_METHOD"],
        "scheme": environ.get("wsgi.url_scheme", "http"),
        "path": environ["PATH_INFO"].encode("latin1").decode("utf8"),
        "query_string": environ["QUERY_STRING"].encode("ascii"),
        "root_path": environ.get("SCRIPT_NAME", "").encode("latin1").decode("utf8"),
        "client": client,
        "server": (environ["SERVER_NAME"], int(environ["SERVER_PORT"])),
        "headers": headers,
        "remote_user": remote_user,
    }
```

**Comment by abersheeran** [comment]
Great idea, but it is not part of the ASGI standard. You should go to [asgiref](https://github.com/django/asgiref) to create a PR. When it becomes a standard, a2wsgi will quickly follow up on this part.

**Comment by abersheeran** [comment]
In addition, perhaps a2wsgi should allow users to access Environ directly from Scope or access Scope from Environ. So all non-standard fields can be accessed, what do you think?

**Reply by userg3003** [reply]
I agree that providing access to environ from the Scope is a great idea. And, in addition, it would solve the problem of non-compliance with the ASGI standard.

**Reply by abersheeran** [reply]
Welcome to create related PR. 😀

**Reply by abersheeran** [reply]
https://github.com/abersheeran/a2wsgi/commit/eecac894aa46a9417b3c2b5df3e5c315ce96f5ee

**Reply by userg3003** [reply]
Thanks, great job.

**Reply by abersheeran** [reply]
This feature has been released, you can try version 1.4.0. If there is any bug, welcome to initiate an ISSUE. 😀
